from snakemake.utils import min_version
from pathlib import Path

##### set minimum snakemake version #####
min_version("8.8.0")

configfile: "config.yaml"

##### resolve global paths from data_dir for input and databases #####
base = Path(config["data_dir"])
config["samples"] = str(base / config["samples"])
for k in ("long_reads", "short_reads", "read_qc", "dehost"):
    config["input"][k] = str(base / config["input"][k])

config["reference"]["human"] = str(Path(config["input"]["databases"]) / config["reference"]["human"])
config["reference"]["bacteria"] = str(Path(config["input"]["databases"]) / config["reference"]["bacteria"])


##### samples #####
with open(config["samples"]) as f:
    next(f)  # skip header
    SAMPLES = [line.split()[0] for line in f if line.strip()]

##### entry point #####
rule all:
    input:
        expand("m/{sample}_binning/bin_refinement/metawrap_70_10_bins.stats", sample=SAMPLES)
#        expand("m/{sample}_binning/concoct_bins/bin.1.fa"), sample=SAMPLES)
#        expand(str(base / config["output"]["assembly"]["binning"] / "{sample}_binning/metabat2_bins/bin.1.fa"), sample=SAMPLES),
#        expand(str(base / config["output"]["assembly"]["binning"] / "{sample}_binning/maxbin2_bins/bin.1.fa"), sample=SAMPLES) 
#        expand(str(base / config["output"]["assembly"]["spades"] / "{sample}_assembly" / "contigs.fasta"), sample=SAMPLES) 
#        expand(str(base / config["output"]["qc"]["read_qc"] / "{sample}" / "final_pure_reads_1.fastq"), sample=SAMPLES)
#        expand(str(base / config["output"]["dehost"]["sam"] / "nano.{sample}.dehost.fq.gz"), sample=SAMPLES)

##### load rules #####
#include: "rules/nano_dehost.smk"
#include: "rules/read_qc.smk"
#include: "rules/preassembly.smk"
include: "rules/binning.smk"


##### utility rules #####
# Create shorter symlinked paths for refined bins (avoids CheckM path-length issues)
rule symlink_binning:
    output:
        "m"
    shell:
        """
        ln -sfn {base}/hybrid_temp/binning {output}
        """

